src/widget/wfindonwebmenu.cpp

```diff
@@ -0,0 +1,178 @@
+    switch (service) {
+    case WFindOnWebMenu::Service::Discogs:
+        return kSearchUrlDiscogsGen;
+    case WFindOnWebMenu::Service::LastFm:
+        if (trackSearchProperty == WFindOnWebMenu::TrackSearchProperties::Title ||
+                trackSearchProperty == WFindOnWebMenu::TrackSearchProperties::ArtistAndTitle) {
+            return kSearchUrlLastFmTitle;
+        } else if (trackSearchProperty == WFindOnWebMenu::TrackSearchProperties::Artist) {
+            return kSearchUrlLastFmArtist;
+        } else {
+            return kSearchUrlLastFmAlbum;
+        }
+    case WFindOnWebMenu::Service::SoundCloud:
+        if (trackSearchProperty == WFindOnWebMenu::TrackSearchProperties::Title ||
+                trackSearchProperty == WFindOnWebMenu::TrackSearchProperties::ArtistAndTitle) {
+            return kSearchUrlSoundCloudTitle;
+        } else if (trackSearchProperty == WFindOnWebMenu::TrackSearchProperties::Artist) {
+            return kSearchUrlSoundCloudArtist;
+        } else {
+            return kSearchUrlSoundCloudAlbum;
+        }
+    default:
+        return kSearchUrlDefault;
+    }
+}
+} // namespace
+
+WFindOnWebMenu::WFindOnWebMenu(QWidget* parent)
+        : QMenu(tr("Find on Web"), parent) {
+}
+
+void WFindOnWebMenu::addSubmenusForServices(const Track& track) {
+    WFindOnWebMenu::populateFromTrackProperties(
+            track,
+            QString(tr("SoundCloud")),
+            Service::SoundCloud);
+    WFindOnWebMenu::populateFromTrackProperties(
+            track, QString(tr("LastFm")), Service::LastFm);
+    WFindOnWebMenu::populateFromTrackProperties(
+            track, QString(tr("Discogs")), Service::Discogs);
+}
+
+void WFindOnWebMenu::addActions(Service service,
+        const QString& trackProperty,
+        QMenu* pServiceMenu,
+        TrackSearchProperties trackSearchProperty) {
+    const QString prefixAction = composePrefixAction(trackSearchProperty);
+    pServiceMenu->addAction(
+            composeActionText(prefixAction, trackProperty),
+            this,
+            [this, service, trackSearchProperty, trackProperty]() {
+                emit triggerBrowser(service, trackSearchProperty, trackProperty);
+            });
+}
+
+void WFindOnWebMenu::openInBrowser(Service service,
+        TrackSearchProperties trackSearchProperty,
+        const QString& query) {
+    const QString serviceUrl = getServiceUrl(service, trackSearchProperty);
+    QUrlQuery urlQuery;
+    urlQuery.addQueryItem("q", query);
+    QUrl url(serviceUrl);
+    url.setQuery(urlQuery);
+    if (!QDesktopServices::openUrl(url)) {
+        qWarning() << "QDesktopServices::openUrl() failed for " << url;
+        DEBUG_ASSERT(false);
+    }
+}
+
+bool WFindOnWebMenu::hasEntriesForTrack(const Track& track) {
+    const auto artist = track.getArtist();
+    const auto trackTitle = track.getTitle();
+    const auto album = track.getAlbum();
+    return (artist.isEmpty() && album.isEmpty() && trackTitle.isEmpty());
```

Comment:
@daschuer If all those properties are empty the menu _has entries_. Sure?

Instead of complaining about the usage of `auto` I recommend to focus on what really matters.

The expression `track.getArtist().isEmpty()` also doesn't reveal the type of the intermediate result and would be permitted according to our rules. Your "the type is out of sight" arguments are invalid.

Reply:
Good catch.